### ISSUE TITLE
IBM User Build - Error uploading copybooks previously downloaded from z/OS-USS

### ISSUE THREAD

**Issue by FALLAI-Denis** [IBM/zopeneditor-about#333]
## Development environment used

- Z Open Editor version: 3.1.0
- Editor Platform
  - [X] Visual Studio Code
  - [ ] Red Hat CodeReady Workspaces
  - [ ] Eclipse Che
  - [ ] Standalone Theia
- Editor Platform Version: 1.74.3
- Operating System: Windows 10 21H2
- Java Version: Java 11
- [ ] Related to RSE API?
  - RSE API Plugin version:
  - Zowe CLI version:
  - Node.js version:
- Logs attached: yes

[Setup_IBM_User_Build.txt](https://github.com/IBM/zopeneditor-about/files/10990309/Setup_IBM_User_Build.txt)

[IBM_User_Build_with_Full_upload.txt](https://github.com/IBM/zopeneditor-about/files/10990291/IBM_User_Build_with_Full_upload.txt)

## Problem Description

Detailed steps for reproducing the problem:

1. Run setup for IBM User Build
2. Run IBM User Build with full upload

### Observed behavior

For "IBM User building" a COBOL program, ZOE detects referenced copybooks and searches for them.
Some copybooks are local to the Workspace (Current Git Repository, or imported Git Repositories), other copybooks are downloaded from the remote z/OS-USS system.

During the total build, ZOE uploads the copybooks to the working folder on z/OS-USS:
* local copybooks are well uploaded
* copybooks retrieved from z/OS-USS are not uploaded correctly: the target path is incomplete and triggers an error in the z/OSMF command

```
2023-03-16T10:51:33.077Z: Fetching build file dependencies...
2023-03-16T10:51:41.326Z: Found a total of 23 local include files:
2023-03-16T10:51:41.327Z: c:\Trv\DepotsGIT\ceab3-environn-outinfr-central-mvp2\.imports\GCL\src\bloc\application\domaine\COPYCOB\SGFDAPP.cpy
2023-03-16T10:51:41.328Z: c:\Trv\DepotsGIT\ceab3-environn-outinfr-central-mvp2\.imports\GCL\src\bloc\application\domaine\COPYCOB\SGFDADM.cpy
2023-03-16T10:51:41.329Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AAA5LSNN.cpy
2023-03-16T10:51:41.330Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AGAVBATC.cpy
2023-03-16T10:51:41.331Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\ADAPDATE.cpy
2023-03-16T10:51:41.333Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\ADADDATE.cpy
2023-03-16T10:51:41.334Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\S9FDI902.cpy
2023-03-16T10:51:41.335Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\S9FDI903.cpy
2023-03-16T10:51:41.336Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AGADPAC0.cpy
2023-03-16T10:51:41.337Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AGAVINIT.cpy
2023-03-16T10:51:41.338Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AGADHORO.cpy
2023-03-16T10:51:41.339Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AAA5LSF0.cpy
2023-03-16T10:51:41.340Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AAAVACCE.cpy
2023-03-16T10:51:41.341Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAAPACCE.cpy
2023-03-16T10:51:41.343Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AGAPAUD1.cpy
2023-03-16T10:51:41.344Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AGAPAUD2.cpy
2023-03-16T10:51:41.345Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAA0LS00.cpy
2023-03-16T10:51:41.346Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAA5LS00.cpy
2023-03-16T10:51:41.347Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAA5L000.cpy
2023-03-16T10:51:41.348Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAA0LSSY.cpy
2023-03-16T10:51:41.350Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAA0L000.cpy
2023-03-16T10:51:41.351Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAADACCE.cpy
2023-03-16T10:51:41.352Z: c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AGADAUDT.cpy
[...]
2023-03-16T10:51:42.301Z: Uploading file c:\Trv\DepotsGIT\ceab3-environn-outinfr-central-mvp2\.imports\GCL\src\bloc\application\domaine\COPYCOB\SGFDAPP.cpy to /u/userTSO/dbb/CEAB3.MVP2/.imports/GCL/src/bloc/application/domaine/COPYCOB/SGFDAPP.cpy with encoding ibm-1147
2023-03-16T10:51:42.322Z: Uploading file c:\Trv\DepotsGIT\ceab3-environn-outinfr-central-mvp2\.imports\GCL\src\bloc\application\domaine\COPYCOB\SGFDADM.cpy to /u/userTSO/dbb/CEAB3.MVP2/.imports/GCL/src/bloc/application/domaine/COPYCOB/SGFDADM.cpy with encoding ibm-1147
2023-03-16T10:51:42.338Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AAA5LSNN.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.351Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AGAVBATC.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.361Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\ADAPDATE.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.373Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\ADADDATE.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.385Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\S9FDI902.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.397Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\S9FDI903.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.407Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AGADPAC0.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.419Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AGAVINIT.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.434Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AGADHORO.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.459Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AAA5LSF0.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.475Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\DSGETU.PERMIBL.SIRIS.COPY\AAAVACCE.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.487Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAAPACCE.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.502Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AGAPAUD1.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.514Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AGAPAUD2.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.526Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAA0LS00.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.536Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAA5LS00.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.548Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAA5L000.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.557Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAA0LSSY.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.568Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAA0L000.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.579Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AAADACCE.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:42.590Z: Uploading file c:\Users\userWindows\.vscodeMainframe\ibm.zopeneditor-3.1.0\resources\temp\_D_\TMVS\RSGREF.PERMIBL.SIRIS.COPY\AGADAUDT.cpy to /u/userTSO/dbb/CEAB3.MVP2 with encoding ibm-1147
2023-03-16T10:51:44.711Z: z/OSMF REST API Error:
Rest API failure with HTTP(S) status 500
category: 6
rc:       8
reason:   91947108
message:  fopen() error
details: 
  - EDC5123I Is a directory. (errno2=0x05620064)


2023-03-16T10:51:44.756Z: z/OSMF REST API Error:
Rest API failure with HTTP(S) status 500
category: 6
rc:       8
reason:   91947108
message:  fopen() error
details: 
  - EDC5123I Is a directory. (errno2=0x05620064)
```

Result of "setup" and "full upload":

```
userTSO:/u/userTSO/dbb/CEAB3.MVP2: >find .
.
./src
./src/SIRIS
./src/SIRIS/COBOL
./src/SIRIS/COBOL/S9TL1A.cbl.build
./src/SIRIS/COBOL/S9TL1A.cbl
./.build
./.build/assemblage.build
./.build/change-E34590.build
./.build/change-T00002.build
./.build/change-T00003.build
./.build/change-xxxxxx.build
./.build/develop.build
./.build/exempleElement.build
./.build/hotfix.build
./.build/master.build
./.build/release-E34590.build
./.build/release-T00003.build
./.build/release-xxxxxx.build
./.build/support-vrX.build
./application-conf
./application-conf/application.properties
./application-conf/Assembler.properties
./application-conf/BMS.properties
./application-conf/Cobol.properties
./application-conf/file.properties
./application-conf/LinkEdit.properties
./.imports
./.imports/GCL
./.imports/GCL/src
./.imports/GCL/src/bloc
./.imports/GCL/src/bloc/application
./.imports/GCL/src/bloc/application/domaine
./.imports/GCL/src/bloc/application/domaine/COPYCOB
./.imports/GCL/src/bloc/application/domaine/COPYCOB/SGFDAPP.cpy
./.imports/GCL/src/bloc/application/domaine/COPYCOB/SGFDADM.cpy
userTSO:/u/userTSO/dbb/CEAB3.MVP2: >
```

**Comment by phaumer** [comment]
This is very odd. User build should only upload local files from your workspace. It must not upload any files retrieved via Zowe from the temp directory (as those file are already on the host and cannot be changed in the editor and could therefore be used for the build from their original locations). Somehow the code that detects those files and excludes them must have gotten a regression. We have to investigate.

**Comment by FALLAI-Denis** [comment]
Hi @phaumer 

That's right, I don't remember ZOE uploading copybooks already coming from the remote system.

Besides, despite these errors, the compilation proceeds normally because the copybooks in error are indeed found in the reference PDS.

That said, it could be of interest to do so: use case where the user build would be done on a machine other than the project builds machine
* ZOE download all copybooks from a remote reference system through zapp file settings
* ZOE upload all local or downloaded copybooks to a remote system dedicated to the user build

This use case could apply to environments managed by Wazi Sandbox:
* the remote reference system, (with IBM DBB), is used for project builds managed by a Jenkins-type orchestrator, with access to copybooks in reference PDS
* the remote test system, (also having IBM DBB), is used for user builds managed by ZOE, not having access to the reference PDS of copybooks (it would be difficult to keep the reference PDS up to date on all instances Wazi Sandbox)

I suggest turning this anomaly into a feature and implementing a setting to decide whether copybooks downloaded from a remote system should be uploaded to the remote user builds system.

**Comment by phaumer** [comment]
We fixed this regression in v3.1.1.
